File: perm/perm.js
function perms(string){
    if (!!string.length && string.length < 2){
        return [string]
    }
    let permArr = [];
    for (let i = 0; i< string.length; i++){
        let charAt = string[i]
        // if (string.indexOf(charAt) !=i){
        //     continue;
        // }
        let rem = string.slice(0,1) + string.slice(i + 1, string.length);

        for (let perm of perms(rem)){
            permArr.push(charAt + perm)
        }
    }
    return permArr;
}

console.log(perms('ab'))
console.log(perms('aabb'))

File: perm/perm.py
def perms(string):
    result = set([string])
    if len(string) == 2:
        result.add(string[1] + string[0])

    elif len(string) > 2:
        for idx, num in enumerate(string):
            for perm in perms(string[:idx] + string[idx+1:]):
                result.add(num + perm)

    return list(result)

print(perms('aabb'))
print(perms('ab'))
print(perms('aab'))